src/internal/connector/onedrive/collection.go

```diff
@@ -223,52 +223,65 @@ func (oc *Collection) populateItems(ctx context.Context) {
 			defer wg.Done()
 			defer func() { <-semaphoreCh }()
 
-			// Read the item
 			var (
+				itemName = *item.GetName()
+				itemSize = *item.GetSize()
 				itemInfo details.ItemInfo
-				itemData io.ReadCloser
-				err      error
-			)
-
-			for i := 1; i <= maxRetries; i++ {
-				itemInfo, itemData, err = oc.itemReader(oc.itemClient, item)
-				if err == nil || graph.IsErrTimeout(err) == nil {
-					// retry on Timeout type errors, break otherwise.
-					break
-				}
-
-				if i < maxRetries {
-					time.Sleep(1 * time.Second)
-				}
-			}
-
-			if err != nil {
-				errUpdater(*item.GetId(), err)
-				return
-			}
-
-			var (
-				itemName string
-				itemSize int64
 			)
 
 			switch oc.source {
 			case SharePointSource:
+				itemInfo.SharePoint = sharePointItemInfo(item, *item.GetSize())
 				itemInfo.SharePoint.ParentPath = parentPathString
-				itemName = itemInfo.SharePoint.ItemName
-				itemSize = itemInfo.SharePoint.Size
 			default:
+				itemInfo.OneDrive = oneDriveItemInfo(item, *item.GetSize())
 				itemInfo.OneDrive.ParentPath = parentPathString
-				itemName = itemInfo.OneDrive.ItemName
-				itemSize = itemInfo.OneDrive.Size
 			}
 
+			// Construct a new lazy readCloser to feed to the collection consumer.
+			// This ensures that downloads won't be attempted unless that consumer
+			// attempts to read bytes.  Assumption is that kopia will check things
+			// like file modtimes before attempting to read.
 			itemReader := lazy.NewLazyReadCloser(func() (io.ReadCloser, error) {
+				// Read the item
+				var (
+					itemData io.ReadCloser
+					err      error
+				)
+
+				for i := 1; i <= maxRetries; i++ {
+					_, itemData, err = oc.itemReader(oc.itemClient, item)
+					if err == nil || graph.IsErrTimeout(err) == nil {
+						break
+					}
+
+					if i < maxRetries {
+						time.Sleep(1 * time.Second)
+					}
+				}
+
+				// check for errors following retries
+				if err != nil {
+					errUpdater(*item.GetId(), err)
+					return nil, err
+				}
+
+				// display/log the item download
 				progReader, closer := observe.ItemProgress(ctx, itemData, observe.ItemBackupMsg, itemName, itemSize)
 				go closer()
+
 				return progReader, nil
 			})
 
+			// This can cause inaccurate counts.  Right now it counts all the items
+			// we intend to read.  Errors within the lazy readCloser will create a
+			// conflict: an item is both successful and erroneous.  But the async
```

Comment:
are the errors propagated to the status updater eventually? Do we need to worry about triggering log messages/failures for the number of items expected not being equal to the sum of the successes and failures in the status updater code?

Reply:
depends on the error; if the lazy reader returns the error, kopia wrapper will catch it.  Yes, if errors occur, counts will not be equal in the status updater.  That's not a blocker atm.